cdc/sink/cdclog/s3.go

```diff
@@ -0,0 +1,592 @@
+		log.Debug("[EmitDDLEvent] list content from s3",
+			zap.String("name", name),
+			zap.Int64("size", size),
+			zap.Any("ddl", ddl))
+	}
+	var fileData []byte
+	if size == 0 || size > maxDDLFlushSize {
+		// no ddl file exists or
+		// exists file is oversized. we should generate a new file
+		fileData = data
+		name = s.prefix + makeDDLFileObject(ddl.CommitTs)
+		log.Debug("[EmitDDLEvent] create first or rotate ddl log",
+			zap.String("name", name), zap.Any("ddl", ddl))
+	} else {
+		// hack way: append data to old file
+		log.Debug("[EmitDDLEvent] append ddl to origin log",
+			zap.String("name", name), zap.Any("ddl", ddl))
+		obj, err := s.getObjectWithKey(ctx, name)
+		if err != nil {
+			return err
+		}
+		buf := make([]byte, maxDDLFlushSize)
+		n, err := obj.Body.Read(buf)
+		if err != nil {
+			if err != io.EOF {
+				return err
+			}
+		}
+		fileData = append(fileData, append(buf[:n], '\n')...)
+	}
+	return s.putObjectWithKey(ctx, name, fileData)
+}
+
+func (s *s3Sink) Initialize(ctx context.Context, tableInfo []*model.SimpleTableInfo) error {
+	if tableInfo != nil {
+		for _, table := range tableInfo {
+			if table != nil {
+				err := s.putObjectWithKey(ctx, s.prefix+makeTableDirectoryName(table.TableID), nil)
+				if err != nil {
+					return errors.Annotate(err, "create table directory on s3 failed")
+				}
+			}
+		}
+		// update log meta to record the relationship about tableName and tableID
+		s.logMeta = makeLogMetaContent(tableInfo)
+
+		data, err := s.logMeta.Marshal()
+		if err != nil {
+			return errors.Annotate(err, "marshal meta to json failed")
+		}
+		return s.putObjectWithKey(ctx, s.prefix+logMetaFile, data)
+	}
+	return nil
+}
+
+func (s *s3Sink) Close() error {
+	return nil
+}
+
+// NewS3Sink creates new sink support log data to s3 directly
+func NewS3Sink(sinkURI *url.URL) (*s3Sink, error) {
+	if len(sinkURI.Host) == 0 {
+		return nil, errors.Errorf("please specify the bucket for s3 in %s", sinkURI)
+	}
+	bucket := sinkURI.Host
+	prefix := strings.Trim(sinkURI.Path, "/")
+	prefix += "/"
+	s3options := new(S3Options)
+	err := s3options.
+		extractFromQuery(sinkURI).
+		adjust()
+	if err != nil {
+		return nil, errors.Annotatef(err, "parse s3 config failed with %s", sinkURI)
+	}
+	awsConfig := s3options.apply()
+	awsSessionOpts := session.Options{Config: *awsConfig}
+	ses, err := session.NewSessionWithOptions(awsSessionOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	s3client := s3.New(ses)
+	err = checkBucket(s3client, bucket)
+	if err != nil {
+		return nil, errors.Errorf("bucket %s is not accessible: %v", bucket, err)
+	}
+
+	notifyChan := make(chan struct{})
+	tableBuffers := make([]*tableBuffer, 0)
+	return &s3Sink{
+		bucket:  bucket,
+		prefix:  prefix,
+		options: s3options,
+		client:  s3client,
+		logMeta: newLogMeta(),
+
+		tableBuffers: tableBuffers,
+		notifyChan:   notifyChan,
+	}, nil
+}
+
+// checkBucket checks if a path exists
+var checkBucket = func(svc *s3.S3, bucket string) error {
```

Comment:
Prefer to declare a function.
```suggestion
func checkBucket(svc *s3.S3, bucket string) error {
```